### PR TITLE
Add `to(stop)` for reading events

### PR DESCRIPTION
**Pull request by anderslemke**
Provide `to(stop)` for reading events until a certain point.

Solves #528.

~~Still missing documentation.~~

### CODE REVIEW

**Comment by anderslemke** [comment]
I think I need some help here... 😞

All but one tests are green, and I have no idea what to do about that last mutation test.

It's about the `BIG_VALUE` being used in `hash`, and to be honest, I don't know what the purpose of that is.

```
 def hash
-  [self.class, get_direction, start, stop, ... , with_types].hash ^ BIG_VALUE
+  [self.class, get_direction, start, stop, ... , with_types].hash
 end
```

Can anyone help me along.
Any pointers would be appreciated.

**Comment by mpraglowski** [comment]
@anderslemke  Looks like you miss a `stop` value in the array here https://github.com/RailsEventStore/rails_event_store/blob/5e086640cc1a0963f55e9034ac8dcd88c04dad0d/ruby_event_store/spec/specification_spec.rb#L554-L564

**Comment by pawelpacana** [comment]
More on that BIG_VALUE: https://blog.arkency.com/relative-testing-vs-absolute-testing/

In short, we don't want https://github.com/RailsEventStore/rails_event_store/blob/5e086640cc1a0963f55e9034ac8dcd88c04dad0d/ruby_event_store/lib/ruby_event_store/specification_result.rb#L204-L217 to equal the hash of an array with same contents. An array is not equal to our value object.

**Comment by pawelpacana** [comment]
btw. once the evil mutation is known, i.e. 
```
evil:RubyEventStore::SpecificationResult#hash:/home/circleci/project/ruby_event_store/lib/ruby_event_store/specification_result.rb:204:4ba2b
```

you can narrow the scope for mutating with:
```
make mutate SUBJECT=RubyEventStore::SpecificationResult#hash
```

**Comment by anderslemke** [comment]
Thank you for your help @mpraglowski and @pawelpacana. 🙏🏼

This PR has definitely been a great learning opportunity for me, where I learned a lot from working with the test suite.

Now that everything is green, I think it's ready for a review.

**Comment by mpraglowski** [comment]
I not sure if `:tail` symbol is best in this use case.

When we have a collection `[1,2,3,4,5]` and we match it with `[H|T]` I would expect `H = 1` and `T = [2,3,4,5]`. In this implementation `:tail` for that collection will be `5`.

Maybe `:end_of_stream` is a better symbol here ? What do you think?

File: ruby_event_store/lib/ruby_event_store/specification.rb [hunk 1]
```diff
@@ -37,6 +37,24 @@ def from(start)
...
+    # {http://railseventstore.org/docs/read/ Find out more}.
+    #
+    # @param start [:tail, String] id of event to start reading from.
+    #   :tail can mean the end or beginning of the stream, depending on the
+    #   #direction
+    # @return [Specification]
+    def to(stop)
+      case stop
+      when Symbol
+        raise InvalidPageStop unless [:tail].include?(stop)
```
[the remark this answers is not in the document]
Comment (anderslemke) [reply ruby_event_store/lib/ruby_event_store/specification.rb]: Yes. It was to keep some kind of symmetry. I can change it 👍

**Comment by anderslemke** [comment]
@mpraglowski That's a very good point! I agree that `:tail` is not good.

`:end_of_stream` could work, even though it is missing some of the purity that `:head` has.

I'm wondering if just using `:end` would work?
Of course `end` is a reserved word in Ruby, and that would probably give reason to some confusion.
But wouldn't we just need the symbol `:end` and the method `end?()`?

Other options could be `:rear`, `:finish`, `:back`, `:tip` or `:last`.

My favourite right now, is `:last`.

**Comment by mpraglowski** [comment]
> My favourite right now, is :last.

👍 better than `:tail` IMHO
Maybe we should have also ':first` as a synonym for `:head` ?
@pawelpacana what do you think?

**Comment by pawelpacana** [comment]
> My favourite right now, is :last.

Let's proceed with that.

> Maybe we should have also :first as a synonym for :head ?

That'd be consistent :+1:

File: railseventstore.org/source/docs/exceptions.html.md [hunk 2]
```diff
@@ -75,6 +75,16 @@ expect do
 end.to raise_error(InvalidPageStart)
 ```
 
+### RubyEventStore::InvalidPageStop
```
Comment (pawelpacana) [inline railseventstore.org/source/docs/exceptions.html.md]: Big ups for the docs! Ideally we want them to be visible after this code is released into a gem. 

We either mark those not-released fragments with annotation that this is unreleased feature or wrap docs code in feature toggle: https://github.com/RailsEventStore/rails_event_store/blob/master/railseventstore.org/config.rb#L23-L41. Example usage: f869949de5298b78e94ca3b1e99a9f6f7bf0895b

**Comment by anderslemke** [comment]
I've tried renaming `:tail` to `:last`. It works, but it's not perfect.
We have a clash with the concept of only reading the last event.
And would have the same clash with `:first`.

As the method `last?` was taken for specifying that you're only interested in the last event, I had to introduce `read_to_last?`. That is clearly missing some symmetry with `head?`.

I'm not sure what I think here, but I'm leaning towards _not_ using the words `first` and `last` for indicating start and stop conditions, and reserving them for getting first and last event only.

That would leave us with choosing another concept for the stop condition.

My actual favourite was `:end`, so I gave that a spin. Let me know what you think.

File: ruby_event_store/lib/ruby_event_store/specification.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (anderslemke) [reply ruby_event_store/lib/ruby_event_store/specification.rb]: I had to use `equal?`, otherwise mutation specs failed.

**Comment by mpraglowski** [comment]
Hmm naming is hard ;) 

This is code where we use then `end?` method from `SpecificationResult` (the one that will clash with another if we go for `:last` name)

```
stream = stream.where(start_condition(spec)) unless spec.head?
stream = stream.where(stop_condition(spec)) unless spec.end?
```

But this is not clear for me what is `spec.head?` and `spec.end?`. Maybe we could replace it with 

```
stream = stream.where(start_condition(spec)) unless spec.start_from?(:head)
stream = stream.where(stop_condition(spec)) unless spec.stop_at?(:end)
```

and then we could go with `:first` & `:last` symbols?

Also in that case it will match the `last?` & `first?` methods meaning.

**Comment by anderslemke** [comment]
I'm not completely sure I understand your point @mpraglowski ... 😬

Are we also going for renaming `:head` and `head?`? If so, it's a different naming game...

Something feels a bit wrong with using `:first` and `:last` for `from()` and `to()`.
When you call `from('abcd')` you get all events **excluding** the event with id `abcd` and all prior events.
When you call `from(:first)` you get all events **including** the first event.

So we need concepts/names for what is **before** the first element, and what is **after** the last element.

In regex it would be `^` and `$`.

Maybe `:beginning` and `:end`?

I'm not sure I think it's way better than `:head`, though...

**Comment by mpraglowski** [comment]
@anderslemke You are right with different meaning of first/last and head/tail - I will merge this PR and create new one to rename head to begin. I like the comparison with regex ^ & $ :) 

Thanks for this PR :)

> So we need concepts/names for what is before the first element, and what is after the last element.

This should be clarified in docs.

**Comment by mpraglowski** [comment]
@anderslemke follow up https://github.com/RailsEventStore/rails_event_store/pull/545

**Pull request merged**